src/AutoRest.CSharp/Common/Output/PostProcessing/PostProcessor.cs

```diff
@@ -0,0 +1,259 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoRest.CSharp.AutoRest.Plugins;
+using AutoRest.CSharp.Mgmt.Output;
+using AutoRest.CSharp.Output.Models.Types;
+using AutoRest.CSharp.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace AutoRest.CSharp.Common.Output.PostProcessing;
+
+internal abstract class PostProcessor
+{
+    protected Project _project;
+
+    public PostProcessor(Project project)
+    {
+        _project = project;
+    }
+
+    protected record ModelSymbols(HashSet<INamedTypeSymbol> DeclaredSymbols, Dictionary<INamedTypeSymbol, HashSet<BaseTypeDeclarationSyntax>> DeclaredNodesCache, Dictionary<Document, HashSet<INamedTypeSymbol>> DocumentCache);
+
+    protected virtual async Task<ModelSymbols> GetModelSymbolsAsync(Compilation compilation, bool publicOnly = true)
+    {
+        var result = new HashSet<INamedTypeSymbol>(SymbolEqualityComparer.Default);
+        var declarationCache = new Dictionary<INamedTypeSymbol, HashSet<BaseTypeDeclarationSyntax>>(SymbolEqualityComparer.Default);
+        var documentCache = new Dictionary<Document, HashSet<INamedTypeSymbol>>();
+        foreach (var document in _project.Documents)
+        {
+            if (!GeneratedCodeWorkspace.IsSharedDocument(document))
+            {
+                var root = await document.GetSyntaxRootAsync();
+                if (root == null)
+                    continue;
+
+                var semanticModel = compilation.GetSemanticModel(root.SyntaxTree);
+
+                foreach (var typeDeclaration in root.DescendantNodes().OfType<BaseTypeDeclarationSyntax>())
+                {
+                    var symbol = semanticModel.GetDeclaredSymbol(typeDeclaration);
+                    if (symbol == null)
+                        continue;
+                    if (publicOnly && symbol.DeclaredAccessibility != Accessibility.Public)
+                        continue;
+                    result.Add(symbol);
+                    declarationCache.AddInList(symbol, typeDeclaration);
+                    documentCache.AddInList(document, symbol, () => new HashSet<INamedTypeSymbol>(SymbolEqualityComparer.Default));
+                }
+            }
+        }
+
+        return new ModelSymbols(result, declarationCache, documentCache);
+    }
+
+    public async Task<Project> InternalizeAsync()
+    {
+        var compilation = await _project.GetCompilationAsync();
+        if (compilation == null)
+            return _project;
+
+        // first get all the declared symbols
+        var definitions = await GetModelSymbolsAsync(compilation, true);
```

Comment:
Is it better to use the real type instead of var here?

Reply:
We are using `var` all over the place. I prefer not changing this.